File: python/1282.py
# Link: https://leetcode.com/problems/group-the-people-given-the-group-size-they-belong-to/description/
# Runtime: 71 ms
# Memory: 16.61 MB

class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        group_size_and_people = defaultdict(lambda: [])
        for index, value in enumerate(groupSizes):
            group_size_and_people[value].append(index)
        
        groups = []
        for group_size, people in group_size_and_people.items():
            while people:
                groups.append(people[:group_size])
                del people[:group_size]
        
        return groups

File: c++/1282.cpp
// Link: https://leetcode.com/problems/group-the-people-given-the-group-size-they-belong-to/description/
// Runtime: 0 ms
// Memory: 13 MB

class Solution 
{
public:
    std::vector<std::vector<int>> groupThePeople(const std::vector<int>& groupSizes) 
    {
        std::unordered_map<int, std::vector<int>> groupSizeAndPeople;
        for (int i = 0; i < static_cast<int>(groupSizes.size()); i++)
            groupSizeAndPeople[groupSizes[i]].push_back(i);
        
        std::vector<std::vector<int>> groups;
        for (auto [key, value] : groupSizeAndPeople)
            while (!value.empty())
            {
                groups.push_back({value.begin(), value.begin() + key});
                value.erase(value.begin(), value.begin() + key);
            }

        return groups;
    }
};
